### PR TITLE
implement ECHO maneuvers for dropships/warships

### PR DESCRIPTION
**Pull request by NickAragua**
This implements the ECHO rules on stratops page 113, which basically allows dropships and warships to "torso twist", at the cost of taking a penalty on their attack rolls.

### CODE REVIEW

File: megamek/src/megamek/common/Dropship.java [hunk 1]
```diff
@@ -1894,5 +1894,64 @@ public void addBattleForceSpecialAbilities(Map<BattleForceSPA,Integer> specialAb
...
+        final int rotate = (dir + (6 - getFacing())) % 6;
+
+        return rotate >= 3 ? (getFacing() + 5) % 6 : (getFacing() + 1) % 6;
+    }
+    
+    @Override
+    public void newRound(int roundNumber) {
+    	super.newRound(roundNumber);
+    	
+    	if(getGame().useVectorMove()) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/Dropship.java:1941]:
```suggestion
    	if (getGame().useVectorMove()) {
```

File: megamek/src/megamek/common/Dropship.java [hunk 2]
```diff
@@ -1894,5 +1894,64 @@ public void addBattleForceSpecialAbilities(Map<BattleForceSPA,Integer> specialAb
...
+    }
+    
+    /**
+     * Utility function that handles situations where a facing change
+     * has some kind of permanent effect on the entity.
+     */
+    @Override
+    public void postProcessFacingChange() {
+    	mpUsed += 2;
+    }
```
Comment (Windchild292) [inline megamek/src/megamek/common/Dropship.java:1955]: This formatting looks weird. Ah, this is tab spaced which should be fixed.

File: megamek/src/megamek/common/actions/WeaponAttackAction.java [hunk 3]
```diff
@@ -3748,6 +3749,19 @@ private static ToHitData compileAeroAttackerToHitMods(IGame game, Entity ae, Tar
                             || wtype.getAtClass() == WeaponType.CLASS_TELE_MISSILE))) {
                 toHit.addModifier(+2, Messages.getString("WeaponAttackAction.AeEvading"));
             }
+            
+            // stratops page 113: ECHO maneuvers for large craft
+            if ((aero instanceof Warship || aero instanceof Dropship) &&
+            		aero.getFacing() != aero.getSecondaryFacing()) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/actions/WeaponAttackAction.java:3755]:
```suggestion
            if (((aero instanceof Warship) || (aero instanceof Dropship)) &&
            		(aero.getFacing() != aero.getSecondaryFacing())) {
```

File: megamek/src/megamek/common/actions/WeaponAttackAction.java [hunk 4]
```diff
@@ -3748,6 +3749,19 @@ private static ToHitData compileAeroAttackerToHitMods(IGame game, Entity ae, Tar
...
+            	// if we're computing this for an "attack preview", then we add 2 MP to 
+            	// the mp used, as we haven't used the MP yet. If we're actually processing
+            	// the attack, then the entity will be marked as 'done' and we have already added
+            	// the 2 MP, so we don't need to double-count it
+            	int extraMP = aero.isDone() ? 0 : 2;
+            	boolean willUseRunMP = aero.mpUsed + extraMP > aero.getWalkMP();
+            	int mod = willUseRunMP ? 2 : 1;
+            	toHit.addModifier(mod, Messages.getString("WeaponAttackAction.LargeCraftEcho"));
+            }
```
Comment (Windchild292) [inline megamek/src/megamek/common/actions/WeaponAttackAction.java:3765]: Tab spacing needs to be fixed.

File: megamek/src/megamek/common/Warship.java [hunk 5]
```diff
@@ -565,4 +565,66 @@ public String getAlphaStrikeLocationName(int index) {
...
+    }
+    
+    /**
+     * Utility function that handles situations where a facing change
+     * has some kind of permanent effect on the entity.
+     */
+    @Override
+    public void postProcessFacingChange() {
+    	mpUsed += 2;
+    }
```
Comment (Windchild292) [inline megamek/src/megamek/common/Warship.java:629]: tab spacing

File: megamek/src/megamek/common/Warship.java [hunk 6]
```diff
@@ -565,4 +565,66 @@ public String getAlphaStrikeLocationName(int index) {
...
+    
+    /**
+     * Handler for when the entity enters a new round
+     */
+    @Override
+    public void newRound(int roundNumber) {
+    	super.newRound(roundNumber);
+    	
+    	if(getGame().useVectorMove()) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/Warship.java:615]:
```suggestion
    	if (getGame().useVectorMove()) {
```

**Comment by NickAragua** [comment]
Heh, different development environment, forgot to configure it the right way.

**Review by Windchild292 — APPROVED** [review]
I've gone through this a few times now and I think this is good to go, but I'm not certain on the rules.

**Pull request merged**